### PR TITLE
tests: click tests for basic and advanced slip39

### LINKED ISSUE #547: Add 1 of 1 Shamir device test
Also consider making a test for max values.

### PR DESCRIPTION
**Pull request by ciny**
This PR resolves #547 and #557 

I'm not entirely sure about the 16 groups/shares advanced test, it takes about 3 minutes on my machine and 5 on CI.

### CODE REVIEW

File: tests/buttons.py [hunk 1]
```diff
@@ -21,8 +21,14 @@ def grid(dim, grid_cells, cell):
 
 CONFIRM_WORD = (MID, TOP)
 
-MINUS = (LEFT, grid(DISPLAY_HEIGHT, 5, 2))
-PLUS = (RIGHT, grid(DISPLAY_HEIGHT, 5, 2))
+MINUS = (LEFT, grid(DISPLAY_HEIGHT, 5, 1))
+PLUS = (RIGHT, grid(DISPLAY_HEIGHT, 5, 1))
```
Comment (tsusanka) [inline tests/buttons.py]: Maybe `RESET_MINUS`? Because these are not used elsewhere, right?

File: tests/click_tests/reset.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+def set_selection(debug, button, diff):
+    layout = debug.read_layout()
+    assert layout.text.startswith("Slip39NumInput")
+    for _ in range(diff):
+        debug.click(button, wait=False)
+    layout = debug.click(buttons.OK, wait=True)
+
+
+def words_read(debug, is_advanced=False):
+    # read words
```
Comment (tsusanka) [inline tests/click_tests/reset.py]: Maybe we can just call the function `read_words`? :)

File: tests/click_tests/reset.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+from trezorlib import messages
+
+from .. import buttons
+
+
+def confirm_wait(debug, startswith):
+    layout = debug.wait_layout()
+    assert layout.text.startswith(startswith)
+    layout = debug.click(buttons.OK, wait=True)
```
Comment (tsusanka) [inline tests/click_tests/reset.py]: You are not using the variable`layout`.

File: tests/click_tests/reset.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+
+def confirm_read(debug, startswith):
+    layout = debug.read_layout()
+    assert layout.text.startswith(startswith)
+    layout = debug.click(buttons.OK, wait=True)
```
Comment (tsusanka) [inline tests/click_tests/reset.py]: dtto

File: tests/click_tests/reset.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+
+def set_selection(debug, button, diff):
+    layout = debug.read_layout()
+    assert layout.text.startswith("Slip39NumInput")
+    for _ in range(diff):
+        debug.click(button, wait=False)
+    layout = debug.click(buttons.OK, wait=True)
```
Comment (tsusanka) [inline tests/click_tests/reset.py]: dtto

File: tests/click_tests/reset.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+def words_read(debug, is_advanced=False):
+    # read words
+    words = []
+    layout = debug.read_layout()
+    if is_advanced:
+        assert layout.text.startswith("Group")
+    else:
+        assert layout.text.startswith("Recovery share")
+    for i in range(6):
+        layout = debug.read_layout()
```
Comment (tsusanka) [inline tests/click_tests/reset.py]: I think this is unused?

File: tests/click_tests/reset.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        if i == 0:
+            words.append(debug.state().layout_lines[3].split()[1])
```
Comment (tsusanka) [inline tests/click_tests/reset.py]: Maybe let's do `lines = debug.state().layout_lines` above the `if` statement?

File: tests/click_tests/reset.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+            words.append(debug.state().layout_lines[4].split()[1])
+            debug.input(swipe=messages.DebugSwipeDirection.UP, wait=True)
+    debug.press_yes()
+
+    return words
+
+
+def words_confirm(debug, words):
+    # confirm words
```
Comment (tsusanka) [inline tests/click_tests/reset.py]: dtto

**Comment by tsusanka** [comment]
Tested ACK, with some nit-picks.

> I'm not entirely sure about the 16 groups/shares advanced test, it takes about 3 minutes on my machine and 5 on CI.

You are correct that the tests take some time, but as we have discussed slow tests are not a reason to omit them. Currently the bottleneck on CI are the device tests, which take about 13 minutes. Thes click tests take around 5 as you have mentioned and run in parallel to the device tests so this is not a problem at the moment.

In future we might consider adding a flag "slow-test" or similar and those would run only on master / over night or something similar (this could lead to broken master though, so discussion on this is needed). But for now, I'd keep it as it is.

File: tests/click_tests/reset.py (hunk 6 above)
Comment (ciny) [reply tests/click_tests/reset.py]: we need to call layout.read_layout() to "refresh" the state so we get the next page. at least that's how it seems to work. @matejcik ?

File: tests/click_tests/reset.py (hunk 2 above)
Comment (ciny) [reply tests/click_tests/reset.py]: ok I renamed it, I also renamed words_confirm to confirm_words to stay consistent.

File: tests/click_tests/reset.py (hunk 6 above)
Comment (ciny) [reply tests/click_tests/reset.py]: but I removed the assignment since we don't use that

Comment (tsusanka) [reply tests/click_tests/reset.py]: That's probably correct, but what I meant is that you do not have to assign it to `layout`.

File: tests/click_tests/reset.py (hunk 7 above)
Comment (tsusanka) [reply tests/click_tests/reset.py]: Also, why not `debug.read_layout()`?

File: tests/click_tests/reset.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+def read_words(debug, is_advanced=False):
+    # read words
+    words = []
+    layout = debug.read_layout()
+    if is_advanced:
+        assert layout.text.startswith("Group")
+    else:
+        assert layout.text.startswith("Recovery share")
+    for i in range(6):
+        debug.read_layout()
```
Comment (tsusanka) [inline tests/click_tests/reset.py]: Btw I guess this returns some text, doesn't it? So we could assert that.

File: tests/click_tests/reset.py [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+        lines = debug.state().layout_lines
```
Comment (tsusanka) [inline tests/click_tests/reset.py]: Is there any reason not to call debug.read_layout?

File: tests/click_tests/reset.py [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+def set_selection(debug, button, diff):
+    layout = debug.read_layout()
+    assert layout.text.startswith("Slip39NumInput")
+    for _ in range(diff):
+        debug.click(button, wait=False)
+    debug.click(buttons.OK, wait=True)
+
+
+def read_words(debug, is_advanced=False):
+    # read words
```
Comment (tsusanka) [inline tests/click_tests/reset.py]: I don't think we need that comment anymore.

File: tests/click_tests/reset.py (hunk 10 above)
Comment (ciny) [reply tests/click_tests/reset.py]: you are right, we can use debug.read_layout().lines to get the same result, fixed

File: tests/click_tests/test_reset_slip39_advanced.py [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+@pytest.mark.skip_t1
+@pytest.mark.setup_client(uninitialized=True)
+def test_reset_slip39_advanced_16of16groups_16of16shares(device_handler):
+    features = device_handler.features()
+    debug = device_handler.debuglink()
+
+    assert features.initialized is False
+
+    os_urandom = mock.Mock(return_value=EXTERNAL_ENTROPY)
+    with mock.patch("os.urandom", os_urandom), device_handler:
```
Comment (matejcik) [inline tests/click_tests/test_reset_slip39_advanced.py]: when you use `device_handler` as a fixture, you don't need to put it in a with-statement

File: tests/click_tests/reset.py (hunk 6 above)
Comment (matejcik) [reply tests/click_tests/reset.py]: You only need to call `read_layout` if you want to use its result. There are no side-effects.
(the result of `read_layout` is a snapshot at the time `read_layout` was called, so you need to call it again to get a new snapshot)

(`wait_layout` on the other hand blocks the test until the layout changes)

File: tests/click_tests/test_reset_slip39_advanced.py [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+        device_handler.run(
+            device.reset,
+            strength=128,
```
Comment (matejcik) [inline tests/click_tests/test_reset_slip39_advanced.py]: this parameter should not be necessary

**Pull request merged**